certora/specs/ERC721.spec

```diff
@@ -0,0 +1,586 @@
+│ Invariant: ownedTotal is the sum of all balances                                                                   │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+invariant ownedTotalIsSumOfBalances()
+    ownedTotal() == sumOfBalances()
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Invariant: balanceOf is the number of token owned                                                                   │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+invariant balanceOfConsistency(address user)
+    balanceOf(user) == ownedByUser[user] &&
+    balanceOf(user) == ghostBalanceOf[user]
+    {
+        preserved {
+            require balanceLimited(user);
+        }
+    }
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Invariant: owner of a token must have some balance                                                                  │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+invariant ownerHasBalance(uint256 tokenId)
+    balanceOf(ownerOf(tokenId)) > 0
+    {
+        preserved {
+            requireInvariant balanceOfConsistency(ownerOf(tokenId));
+            require balanceLimited(ownerOf(tokenId));
+        }
+    }
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Invariant: tokens that do not exist are not owned and not approved                                                  │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+invariant notMintedUnset(uint256 tokenId)
+    (!tokenExists(tokenId) <=> unsafeOwnerOf(tokenId) == 0) &&
+    (!tokenExists(tokenId) => unsafeGetApproved(tokenId) == 0)
+
+/*
+┌─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┐
+│ Rule: ownerOf and getApproved revert if token does not exist                                                        │
+└─────────────────────────────────────────────────────────────────────────────────────────────────────────────────────┘
+*/
+rule notMintedRevert(uint256 tokenId) {
+    bool exist = tokenExists(tokenId);
+
+    address owner = ownerOf@withrevert(tokenId);
+    assert exist <=> !lastReverted;
+    assert exist => owner == unsafeOwnerOf(tokenId); // notMintedUnset tels us this is non-zero
+
+    address approved = getApproved@withrevert(tokenId);
+    assert exist <=> !lastReverted;
+    assert exist => approved == unsafeGetApproved(tokenId);
```

Comment:
again, because `getApproved@withrevert` returns a wildcard on reverting calls, this property is not valid.

If the token doesn't exist:
-  `e` is false
- `approved` is *
- `unsafeGetApproved(tokenId)` is 0.

remeber that `*==0` is true.

Reply:
Clear, thanks!